addons/bus/static/src/services/worker_service.js

```diff
@@ -0,0 +1,133 @@
+import { browser } from "@web/core/browser/browser";
+import { registry } from "@web/core/registry";
+import { Deferred } from "@web/core/utils/concurrency";
+import { session } from "@web/session";
+import { isIosApp } from "@web/core/browser/feature_detection";
+
+export const WORKER_STATE = Object.freeze({
+    UNINITIALIZED: "UNINITIALIZED",
+    INITIALIZING: "INITIALIZING",
+    INITIALIZED: "INITIALIZED",
+    FAILED: "FAILED",
+});
+
+export const workerService = {
+    dependencies: ["bus.parameters"],
+
+    start(env, { "bus.parameters": params }) {
+        let worker;
+        let isUsingSharedWorker = browser.SharedWorker && !isIosApp();
+        let state = WORKER_STATE.UNINITIALIZED;
+        let connectionInitializedDeferred = null;
+
+        function startWorker() {
+            if (!connectionInitializedDeferred) {
+                connectionInitializedDeferred = new Deferred();
+            }
+            state = WORKER_STATE.INITIALIZING;
+            let workerURL = `${params.serverURL}/bus/websocket_worker_bundle?v=${session.websocket_worker_version}`;
+            if (params.serverURL !== window.origin) {
+                // Worker service can be loaded from a different origin than the
+                // bundle URL. The Worker expects an URL from this origin, give
+                // it a base64 URL that will then load the bundle via "importScripts"
+                // which allows cross origin.
+                const source = `importScripts("${workerURL}");`;
+                workerURL = "data:application/javascript;base64," + window.btoa(source);
+            }
+            const workerClass = isUsingSharedWorker ? browser.SharedWorker : browser.Worker;
+            worker = new workerClass(workerURL, {
+                name: isUsingSharedWorker ? "odoo:bus_shared_worker" : "odoo:bus_worker",
+            });
+            worker.onerror = onInitError;
+            _registerHandler(handleMessage);
```

Comment:
No need for a dedicated function: it’s simple enough and only used once. Having a separate function just makes the flow harder to follow.

Reply:
We actually uses it twice. One is here, another one in the public method `registerHandler`.

Reply:
Sorry I meant `handleMessage`, not `_registerHandler`.